ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -662,3 +663,17 @@ getPeersFromCurrentLedgerAfterSlot kernel slotNo =
       case ledgerTipSlot st of
         Origin        -> False
         NotOrigin tip -> tip > slotNo
+
+-- | Retrieve the slot number of the immutable tip
+getImmTipSlotNo ::
+     ( IOLike m
+     , UpdateLedger blk
+     )
+  => NodeKernel m addrNTN addrNTC blk
+  -> STM m SlotNo
+getImmTipSlotNo kernel = do
+    immutableLedger <-
+      ledgerState <$> ChainDB.getImmutableLedger (getChainDB kernel)
+    case ledgerTipSlot immutableLedger of
+      Origin -> retry
+      NotOrigin tip -> return tip
```

Comment:
If the node is starting up without any DB, a call to this is going to block until the node starts catching up. Me and @johnalotoski found an issue where the node without local roots nor bootstrap peers, _only_ public roots, would not make progress. I think this happens because when the node is trying to fetch public roots it is going to call `lpGetLatestSlot` in order to decide if it should fetch peers from the ledger or from public configuration. Since this `lpGetLatestSlot` call blocks, the node doesn't make any progress.

So I ask: Is the best way to solve this to return 0 in the `Origin` case instead of retrying?

Reply:
This function is passed directly as `lpGetLatestSlot` in the interface.

So I think the question is ultimately: should `lpGetLatestSlot` be blocking?

If it should not block, then the type of `lpGetLatestSlot` should be `STM (WithOrigin SlotNo)`. In that case, `getImmTipSlotNo` would end with `pure $ ledgerTipSlot immutableLedger`.

Reply:
`WithOrigin` is defined in `cardano-base`: https://github.com/IntersectMBO/cardano-base/blob/31185617fa029db5332892203c37cb01ae248c65/cardano-slotting/src/Cardano/Slotting/Slot.hs#L50-L51

Reply:
Could we skip one step and provide an abstraction to `SlotNo` instead? We would have to map a value from `Origin` anyway. Is this not desired? Previously I believe you just returned a `Maybe` value

Reply:
We agreed on Slack that it should return `WithOrigin SlotNo`, and something closer to the "business logic" can squash that to `SlotNo 0`, with a comment explaining why that is harmless in this case.